mapo/agents/mapo/mapo_model.py

```diff
@@ -6,34 +6,72 @@
+    """Extension of standard TFModel for MAPO.
+
+    Encapsulates all the networks necessary for Model-Aware Policy Optimization.
+    Exposes the necessary API for loss calculation and variable tracking.
+
+    Keyword arguments:
+        target_networks(bool): Whether or not to create target actor and critic.
+        twin_q=(bool): Whether or not to create twin Q networks.
+    """
 
     def __init__(
-        self, obs_space, action_space, num_outputs, model_config, name, twin_q=False
+        self,
+        obs_space,
+        action_space,
+        num_outputs,
+        model_config,
+        name,
+        target_networks=False,
+        twin_q=False,
     ):
         # pylint: disable=too-many-arguments,arguments-differ
         prep = get_preprocessor(obs_space)(obs_space)
         # Ignore num_outputs as we don't use a shared state preprocessor
         super().__init__(obs_space, action_space, prep.size, model_config, name)
 
-        self.policy = build_deterministic_policy(
+        models = {}
+        models["policy"] = build_deterministic_policy(
             obs_space, action_space, model_config["custom_options"]["actor"]
         )
-        self.register_variables(self.policy.variables)
-        self.q_net = build_continuous_q_function(
+        models["q_net"] = build_continuous_q_function(
             obs_space, action_space, model_config["custom_options"]["critic"]
         )
-        self.register_variables(self.q_net.variables)
-
         self.twin_q = twin_q
         if twin_q:
-            self.twin_q_net = build_continuous_q_function(
+            models["twin_q_net"] = build_continuous_q_function(
+                obs_space, action_space, model_config["custom_options"]["critic"]
+            )
+
+        models["dynamics"] = GaussianDynamicsModel(
+            obs_space, action_space, **model_config["custom_options"]["dynamics"]
+        )
+        self.models = models
+        self.register_variables(
+            [variable for model in models.values() for variable in model.variables]
+        )
+
+        if target_networks:
+            target_models = {}
+            target_models["policy"] = build_deterministic_policy(
+                obs_space, action_space, model_config["custom_options"]["actor"]
+            )
+            target_models["q_net"] = build_continuous_q_function(
                 obs_space, action_space, model_config["custom_options"]["critic"]
             )
-            self.register_variables(self.twin_q_net.variables)
+            if twin_q:
+                target_models["twin_q_net"] = build_continuous_q_function(
+                    obs_space, action_space, model_config["custom_options"]["critic"]
+                )
+            self.target_models = target_models
+            self.register_variables(
+                [var for model in target_models.values() for var in model.variables]
+            )
 
     @override(TFModelV2)
     def forward(self, input_dict, state, seq_lens):
```

Comment:
Just as of curiosity, why not to use `forward()` to compute the `policy_action`, `sampled_next_state`, and `next_state_value`...?

This way you could simplify `_build_actor_loss(policy, batch_tensors)` in `off_mapo_policy.py` module..

Reply:
As a side effect, it would become easier to test these computation that are currently buried in the `_build_actor_loss` function...

Reply:
Unfortunately `ModelV2`'s `__call__` method requires us to return a single tensor of shape `(None, output_size)`. That's why there's the hack in `__init__` where we call `super().__init__` with a predefined output size, and then return in `forward` the flattened observation. I had to do this because `self.model` is called internally by DynamicTFPolicy (see [here](https://hackmd.io/@angelovt/By-MYD77r#build-model-output)).

IMO, this is a bad design decision by the developers, as I would also love to do something like:
```python
policy_action, sampled_next_state, next_state_value = self.model(inputs)
``` 
The only workaround I found was to separate the outputs in different methods, as I didn't want to group everything in one tensor and split it afterwards. We can discuss this if you think it's a good idea though.

Reply:
Maybe a simpler solution would be to compute everything in forward, but only return the policy_action... the other tensors you could plug as attributes of the model ...

just as in the simple example in the documentation below in which the value function tensor is set to `self._value_out` ..

https://ray.readthedocs.io/en/latest/rllib-models.html?highlight=TFModelV2#ray.rllib.models.tf.tf_modelv2.TFModelV2.forward

What do you think of this solution? Would it be worth it so to make the code more testable?

Reply:
I find very hacky since it creates side-effects (changes the model's attributes) after every forward pass, which I'm not a fan of. I think we can still test the loss computation using the custom methods I created. 

Shall we revisit this issue later? (if the need arises)

Reply:
sure! very tangential issue here... nothing to be done right now! thanks